include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -332,39 +428,115 @@ create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_t
 
     Assert(fine_triangulation, ExcNotImplemented());
 
-    // clone distributed triangulation for coarsening
-    parallel::distributed::Triangulation<dim, spacedim> temp_tria(
+    // clone distributed triangulation and start coarsening
+    parallel::distributed::Triangulation<dim, spacedim> tria_copy(
       fine_triangulation->get_communicator(), fine_triangulation->get_mesh_smoothing());
 
-    temp_tria.copy_triangulation(*fine_triangulation);
-
-    // create coarse meshes
-    for(unsigned int l = (fine_triangulation->n_global_levels() - 1); l > 0; --l)
+    tria_copy.copy_triangulation(*fine_triangulation);
+    tria_copy.coarsen_global();
+
+    MPI_Comm mpi_comm = fine_triangulation->get_communicator();
+
+    // as long as we have enough cells per process, we can perform regular
+    // coarsening with all MPI processes TODO: The number of 400 cells per MPI
+    // process (or 50 if the next refinement were done in 3D) was found to be a
+    // good tradeoff between communication cost and workload size of linear
+    // polynomials, resulting in small run times in preliminary studies. This
+    // could be generalized by a parameter to set in the application files.
+    unsigned int n_cells_per_process = 400;
+    for(unsigned int level = fine_triangulation->n_global_levels() - 1;
+        level > 0 && tria_copy.n_global_active_cells() / Utilities::MPI::n_mpi_processes(mpi_comm) >
+                       n_cells_per_process;
+        --level)
     {
-      // coarsen mesh
-      temp_tria.coarsen_global();
+      // extract relevant information from distributed triangulation
+      auto const construction_data =
+        TriangulationDescription::Utilities::create_description_from_triangulation(tria_copy,
+                                                                                   mpi_comm);
 
-      // create empty (fully distributed) triangulation
-      auto new_tria = std::make_shared<parallel::fullydistributed::Triangulation<dim, spacedim>>(
-        fine_triangulation->get_communicator());
+      // create fully distributed triangulation
+      auto level_tria =
+        std::make_shared<parallel::fullydistributed::Triangulation<dim, spacedim>>(mpi_comm);
 
       for(auto const i : fine_triangulation->get_manifold_ids())
         if(i != numbers::flat_manifold_id)
-          new_tria->set_manifold(i, fine_triangulation->get_manifold(i));
+          level_tria->set_manifold(i, fine_triangulation->get_manifold(i));
+
+      level_tria->create_triangulation(construction_data);
+
+      coarse_grid_triangulations[level - 1] = level_tria;
+
+      if(level > 1)
```

Comment:
In my opinion, the state of `level` does not fit to the coarsening state of `tria_copy`, which makes the code difficult to understand.

Since `tria_copy` has already been coarsened in line 436, the initialization of `level` in the for-loop should be different. Here in this line, this change would imply `if(level > 0)`. I will add suggestions to the other places where `level` is used.